### DISCUSSION TITLE
ESP32 device  dos not show up after disconnect

### DISCUSSION THREAD

**Discussion by DrWino** [h2zero/NimBLE-Arduino#347] [General]
I am having problem using ESP32-arduino. I used the original BLE library and I can connect with Android. But if I disconnect, the device no longer shows up on the BT list. The only way is to reboot the ESP32 and then it show up again. Obviously not useful. I saw some discussion about this problem on the net but nonee of the solutions worked for me. So I decided to try the NimBLE library and used the migration hints to modify BLE server. It works BUT the same problem. Once you disconnect it does not show up again. What gives?

**Comment by h2zero** [comment]
You need to start advertising again when the client disconnects. You can do this manually in the `onDisconnect()` callback or you can enable automatic advertising on disconnect in setup using `NimBLEServer::advertiseOnDisconnect(true);`

**Reply by DrWino** [reply]
Thanks for the response but I could NOT make it work - I include nimBLEServer.h in addition to nimBLEDevice.h - 

  BLEDevice::init("GOCOLOR");
  BLEServer *pServer = BLEDevice::createServer();
  BLEService *pService = pServer->createService(SERVICE_UUID);
  BLECharacteristic *pCharacteristic = pService->createCharacteristic(
                                         CHARACTERISTIC_UUID,
                                         NIMBLE_PROPERTY::READ |
                                         NIMBLE_PROPERTY::WRITE 
                                       );

  pCharacteristic->setCallbacks(new MyCallbacks());
  pService->start();
  BLEAdvertising *pAdvertising = BLEDevice::getAdvertising();
  pServer->advertiseOnDisconnect(true);
  pAdvertising->addServiceUUID(SERVICE_UUID);
  pAdvertising->setScanResponse(true);
  pAdvertising->setMinPreferred(0x06);  // functions that help with iPhone connections issue
  pAdvertising->setMinPreferred(0x12);
  BLEDevice::startAdvertising();
  pServer->advertiseOnDisconnect(true);

Device does not show up after first disconnect. I am doing something stupid! Not sure how to add the OnDisconnect call back.

**Reply by h2zero** [reply]
What are you using to scan for devices? I cant reproduce the problem using nRFconnect.

**Reply by skysingh** [reply]
Works now. I could not get the automatic mode to work but I was able to do it via onDisconnect() callback. Thanks for the nice library and support.
